File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) <= 2:
            return max(nums)
        dp = [nums[0], max(nums[0], nums[1])]

        for i in range(2, len(nums)):
            dp.append(max(dp[-1], nums[i] + dp[-2]))

        return dp[-1]

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() >= 2) nums[1] = max(nums[0], nums[1]);
        else return nums.back();
        for (int i = 2; i < nums.size(); ++i) {
            nums[i] = max(nums[i-2] + nums[i], nums[i-1]);
        }
        return nums.back();
    }
};